src/main/java/net/minecraftforge/fml/common/registry/FMLControlledNamespacedRegistry.java

```diff
@@ -766,11 +766,102 @@ public void notifyCallbacks()
     @Override
     public ResourceLocation getNameForObject(I p_177774_1_)
     {
-        ResourceLocation rl = super.getNameForObject(p_177774_1_);
+        ResourceLocation rl = super.getNameForObjectBypass(p_177774_1_);
         if (rl == null)
         {
             rl = activeSubstitutions.inverse().get(p_177774_1_);
         }
         return rl;
     }
+
+    // IForgeRegistry: Modders should only interfaces with these methods
+    @Override
+    public void register(I value)
+    {
+        ResourceLocation key = value.getRegistryName();
+        if (key == null)
+        {
+            FMLLog.severe("Attempted to register a entry with a null name: %s", value);
+            throw new NullPointerException(String.format("Attempted to register a entry with a null name: %s", value));
+        }
+        add(-1, key, value);
+    }
+
+    @Override
+    public boolean contains(ResourceLocation key)
+    {
+        return containsKey(key);
+    }
+
+    @Override
+    public boolean contains(I value)
+    {
+        return get(value) != null;
+    }
+
+    @Override
```

Comment:
The "get" function can either get the location from the object, or the object from the location, which is 2 differents behaviours. I think it's confusing. Suggestion : keep 2 functions with differents names for each purpose ("getObject" and "getNameForObject" seams fine)